src/clsag/sign.rs

```diff
@@ -0,0 +1,106 @@
+use crate::clsag::{array_map, compute_L, compute_R, INV_EIGHT, RING_SIZE};
+use crate::util::ringct::Clsag;
+use curve25519_dalek::edwards::EdwardsPoint;
+use curve25519_dalek::scalar::Scalar;
+
+/// Signs a message with the given signing key using Monero's implementation of the CLSAG signature algorithm.
+///
+/// # Design notes
+///
+/// This implementations is purposely defined to be pure and low-level in the way that it doesn't make any assumptions about how certain values are generated, even if it is recommended for them to be random (like `fake_responses`).
+/// This serves primarily two purposes:
+///
+/// 1. We can very easily unit test the implementation.
+/// 2. It allows for more esoteric uses of the functionality like adaptor signatures (which need to control how L & R are computed for example).
+#[allow(non_snake_case)]
+#[allow(clippy::too_many_arguments)]
+pub fn sign(
+    msg: &[u8; 32],
+    signing_key: Scalar,
+    signing_key_index: usize,
+    H_p_pk: EdwardsPoint,
+    alpha: Scalar,
+    ring: &[EdwardsPoint; RING_SIZE],
+    commitment_ring: &[EdwardsPoint; RING_SIZE],
+    fake_responses: [Scalar; RING_SIZE - 1],
+    z: Scalar,
+    pseudo_output_commitment: EdwardsPoint,
+    L: EdwardsPoint,
+    R: EdwardsPoint,
+    I: EdwardsPoint,
```

Comment:
It really is a lot of arguments, Clippy! It would be good to document exactly what each of these arguments is, especially if this is to be a public API.

Reply:
Agreed. Recorded a ticket here: https://github.com/comit-network/monero-rs/issues/25